scripts/asr_language_modeling/ngram_lm/eval_beamsearch_ngram_transducer.py

```diff
@@ -13,257 +13,420 @@
-        rnnt_cfg.beam.beam_size = args.beam_width
-        rnnt_cfg.beam.ngram_lm_model = args.kenlm_model_file
-        rnnt_cfg.beam.ngram_lm_alpha = args.beam_alpha  # 0.2, 0.3
-        rnnt_cfg.compute_hypothesis_token_set = False
-        rnnt_cfg.beam.return_best_hypothesis = False
-        rnnt_cfg.beam.maes_prefix_alpha = args.maes_prefix_alpha
-        rnnt_cfg.beam.maes_expansion_gamma = args.maes_expansion_gamma
-        rnnt_cfg.beam.hat_subtract_ilm = args.hat_subtract_ilm
-        rnnt_cfg.beam.hat_ilm_weight = args.hat_ilm_weight
-        asr_model.change_decoding_strategy(OmegaConf.structured(rnnt_cfg))
 
         @contextlib.contextmanager
         def default_autocast():
             yield
 
-        if args.use_amp:
+        if cfg.use_amp:
             if torch.cuda.is_available() and hasattr(torch.cuda, 'amp') and hasattr(torch.cuda.amp, 'autocast'):
                 logging.info("AMP is enabled!\n")
                 autocast = torch.cuda.amp.autocast
+
             else:
                 autocast = default_autocast
         else:
-            autocast = default_autocast
 
-        params = {'beam_width': args.beam_width, 'beam_alpha': args.beam_alpha}
-
-        logging.info(f"==============================Starting the beam search decoding===============================")
-        logging.info(f"Beam search params: {params}")
-        logging.info(f"It may take some time...")
-        logging.info(f"==============================================================================================")
+            autocast = default_autocast
 
         with autocast():
             with torch.no_grad():
-                hypotheses, all_hypotheses = asr_model.transcribe(
-                    audio_file_paths, batch_size=args.acoustic_batch_size, return_hypotheses=True
+                all_logits, _ = asr_model.transcribe(
+                    audio_file_paths, batch_size=cfg.acoustic_batch_size, return_encoder_embiddings=True
                 )
 
-    # delete the model to free the memory
-    del asr_model
+        all_probs = all_logits
+        if cfg.probs_cache_file:
+            logging.info(f"Writing pickle files of probabilities at '{cfg.probs_cache_file}'...")
+            with open(cfg.probs_cache_file, 'wb') as f_dump:
+                pickle.dump(all_probs, f_dump)
+
+    if cfg.decoding_strategy == "greedy_batch":
+        asr_model = asr_model.to('cpu')
+        candidate_wer, candidate_cer = decoding_step(
+            asr_model,
+            cfg,
+            all_probs=all_probs,
+            target_transcripts=target_transcripts,
+            beam_batch_size=cfg.beam_batch_size,
+            progress_bar=True,
+        )
+        logging.info(f"Greedy batch WER/CER = {candidate_wer:.2%}/{candidate_cer:.2%}")
+
+    asr_model = asr_model.to('cpu')
+
+    # 'greedy_batch' decoding_strategy would skip the beam search decoding
+    if cfg.decoding_strategy in ["maes"]:
```

Comment:
RNNT supports all beam search inference algos, do not special case like this to maes just because of ILM. 
This is meant to be a generic inference script of RNNT models with beam search, not special cased external LM inference with HAT alone using maes only.

Reply:
Only maes decoding supports ngram lm fusion for RNNT/HAT now. Do we need to have all beamsearch decoding methods in the `eval_beamsearch_ngram_transducer.py` script even if they do not support ngram lm?

Reply:
Treat Ngram as special optional case. This script should support beam search in general (all RNNT capable beam search algos) and if you're passing external kenlm then enforce that only mAES is currently supported

Reply:
I added support for all the RNNT decoding algorithms.